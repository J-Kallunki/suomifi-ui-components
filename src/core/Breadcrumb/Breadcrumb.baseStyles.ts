import { css } from '@emotion/core';
import { suomifiTheme } from '../theme';
import { BreadcrumbProps } from './Breadcrumb';
import { nav, list, listItem, fonts } from '../theme/reset';

export const baseStyles = ({ theme = suomifiTheme }: BreadcrumbProps) => css`
  ${nav}
  ${fonts(theme).body}
  background-color: ${theme.colors.whiteBase};

  & .fi-breadcrumb {
    &-list {
      ${list}
      ${fonts(theme).body}
      margin: 0;
      padding: 0;
    }
    &-item {
      ${listItem}
      ${fonts(theme).body}
      float: left;
    }
    &-item,
    &-link,
    &-icon {
      font-size: ${theme.typography.fontSize.body};
    }
  }
`;
