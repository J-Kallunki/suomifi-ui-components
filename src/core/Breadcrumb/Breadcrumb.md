```jsx
<Breadcrumb aria-label="breadcrumb">
  <Breadcrumb.link href="/">Frontpage</Breadcrumb.link>
  <Breadcrumb.link href="/sub">Sub page</Breadcrumb.link>
  <Breadcrumb.link current>Sub sub page</Breadcrumb.link>
</Breadcrumb>
```
