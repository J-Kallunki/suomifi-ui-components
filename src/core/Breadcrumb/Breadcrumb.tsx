import React, { Component, Fragment } from 'react';
import styled from '@emotion/styled';
import { withDefaultTheme } from '../theme/utils/defaultTheme';
import { ThemeComponent, ThemeProp } from '../theme';
import { baseStyles } from './Breadcrumb.baseStyles';
import {
  Breadcrumb as CompBreadcrumb,
  BreadcrumbProps as CompBreadcrumbProps,
  baseClassName,
} from '../../components/Breadcrumb/Breadcrumb';
import { Link, LinkProps } from '../Link/Link';
import { Icon } from '../Icon/Icon';
import { HtmlSpan } from '../../reset';
import classnames from 'classnames';

const linkClassName = `${baseClassName}-link`;
const iconClassName = `${baseClassName}-icon`;

type BreadcrumbVariant = 'default' | 'link';

export interface BreadcrumbProps extends CompBreadcrumbProps, ThemeComponent {
  /**
   * 'default' | 'expansion'
   * @default default
   */
  variant?: BreadcrumbVariant;
}

export interface BreadcrumbLinkProps extends LinkProps {
  /** Indicating the link is the current page */
  current?: boolean;
  theme: ThemeProp;
}

const StyledBreadcrumb = styled(({ theme, ...passProps }: BreadcrumbProps) => (
  <CompBreadcrumb {...passProps} />
))`
  ${props => baseStyles(props)};
`;

const BreadcrumbLink = ({
  theme,
  current,
  children,
  className,
  ...passProps
}: BreadcrumbLinkProps) =>
  !!current ? (
    <HtmlSpan className={className} aria-current="page">
      {children}
    </HtmlSpan>
  ) : (
    <Fragment>
      <Link {...passProps} className={classnames(linkClassName, className)}>
        {children}
      </Link>
      <Icon
        icon="linkBreadcrumb"
        className={iconClassName}
        color={theme.colors.text}
      />
    </Fragment>
  );

type VariantBreadcrumbProps =
  | BreadcrumbProps
  | LinkProps & { variant?: BreadcrumbVariant };

/**
 * Used for navigation path
 */
export class Breadcrumb extends Component<VariantBreadcrumbProps> {
  static link = (props: BreadcrumbLinkProps) => {
    return <BreadcrumbLink {...withDefaultTheme(props)} />;
  };

  render() {
    const { variant, 'aria-label': ariaLabel, ...passProps } = withDefaultTheme(
      this.props,
    );
    if (variant === 'link') {
      return <BreadcrumbLink {...passProps as BreadcrumbLinkProps} />;
    }
    return (
      <StyledBreadcrumb
        {...passProps}
        aria-label={!!ariaLabel ? ariaLabel : 'breadcrumb'}
      />
    );
  }
}
