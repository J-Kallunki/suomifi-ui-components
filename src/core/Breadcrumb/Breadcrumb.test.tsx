import React from 'react';
import { render } from 'react-testing-library';

import { Breadcrumb } from './Breadcrumb';

test('calling render with the same component on the same container does not remount', () => {
  const BreadcrumbRendered = render(
    <Breadcrumb data-testid="Breadcrumb" aria-label="breadcrumb">
      <Breadcrumb.link href="/">Frontpage</Breadcrumb.link>
      <Breadcrumb.link href="/sub">Sub page</Breadcrumb.link>
      <Breadcrumb.link current={true}>Sub sub page</Breadcrumb.link>
    </Breadcrumb>,
  );
  const { getByTestId, container } = BreadcrumbRendered;
  expect(container.firstChild).toMatchSnapshot();
  expect(getByTestId('Breadcrumb').textContent).toBe('Test');
});
