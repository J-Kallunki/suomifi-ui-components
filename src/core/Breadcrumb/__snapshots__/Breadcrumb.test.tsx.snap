// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calling render with the same component on the same container does not remount 1`] = `
.emotion-9 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: block;
  max-width: 100%;
  color: #282828;
  display: block;
  font-family: 'Source Sans Pro','Helvetica Neue',Arial;
  font-size: 18px;
  font-weight: 400;
  line-height: 1em;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-font-smoothing: antialiased;
  background-color: #FFFFFF;
}

.emotion-9 .fi-breadcrumb-list {
  color: #282828;
  list-style: none;
  font-family: 'Source Sans Pro','Helvetica Neue',Arial;
  font-size: 18px;
  font-weight: 400;
  line-height: 1em;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-font-smoothing: antialiased;
  margin: 0;
  padding: 0;
}

.emotion-9 .fi-breadcrumb-item {
  color: #282828;
  list-style: none;
  font-family: 'Source Sans Pro','Helvetica Neue',Arial;
  font-size: 18px;
  font-weight: 400;
  line-height: 1em;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-font-smoothing: antialiased;
  float: left;
}

.emotion-9 .fi-breadcrumb-item,
.emotion-9 .fi-breadcrumb-link,
.emotion-9 .fi-breadcrumb-icon {
  font-size: 18px;
}

.emotion-8 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: block;
  list-style-type: decimal;
  margin-top: 1em;
  margin-bottom: 1em;
  margin-left: 0;
  margin-right: 0;
  padding-left: 40px;
}

.emotion-2 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: list-item;
}

.emotion-0 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  background-color: transparent;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  color: #282828;
  font-family: 'Source Sans Pro','Helvetica Neue',Arial;
  font-size: 18px;
  font-weight: 400;
  line-height: 1em;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-font-smoothing: antialiased;
  color: #2A6EBB;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-0:hover,
.emotion-0:active,
.emotion-0:focus,
.emotion-0:focus-within,
.emotion-0:visited {
  color: inherit;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  cursor: pointer;
}

.emotion-0:focus {
  outline: 0;
  position: relative;
}

.emotion-0:focus:after {
  content: '';
  position: absolute;
  top: -4px;
  right: -4px;
  bottom: -4px;
  left: -4px;
  border-radius: 4px;
  background-color: transparent;
  border: 1px solid #E97025;
  box-sizing: border-box;
  box-shadow: 0 0 10px 0 #E97025;
  z-index: 9999;
}

.emotion-0:focus:not(:focus-visible):after {
  content: none;
}

.emotion-0:hover,
.emotion-0:active,
.emotion-0:focus,
.emotion-0:focus-within,
.emotion-0:visited {
  color: #2A6EBB;
}

.emotion-0:hover,
.emotion-0:active {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.emotion-0:visited {
  color: #9F60B5;
}

.emotion-1 {
  display: inline-block;
  vertical-align: bottom;
}

.emotion-6 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

<nav
  aria-label="breadcrumb"
  class="fi-breadcrumb emotion-9"
  data-testid="Breadcrumb"
>
  <ol
    class="fi-breadcrumb-list emotion-8"
  >
    <li
      class="fi-breadcrumb-item emotion-2"
    >
      <a
        class="fi-breadcrumb-link fi-link emotion-0"
        href="/"
      >
        Frontpage
      </a>
      <svg
        aria-hidden="true"
        class="fi-breadcrumb-icon emotion-1"
        focusable="false"
        height="1em"
        style="fill: #282828;"
        viewBox="0 0 24 24"
        width="1em"
      >
        <path
          d="M6 18v-2.12l5.902-2.328 3.748-1.492v-.12l-3.748-1.492L6 8.119V6l12 5.015v1.97z"
        />
      </svg>
    </li>
    <li
      class="fi-breadcrumb-item emotion-2"
    >
      <a
        class="fi-breadcrumb-link fi-link emotion-0"
        href="/sub"
      >
        Sub page
      </a>
      <svg
        aria-hidden="true"
        class="fi-breadcrumb-icon emotion-1"
        focusable="false"
        height="1em"
        style="fill: #282828;"
        viewBox="0 0 24 24"
        width="1em"
      >
        <path
          d="M6 18v-2.12l5.902-2.328 3.748-1.492v-.12l-3.748-1.492L6 8.119V6l12 5.015v1.97z"
        />
      </svg>
    </li>
    <li
      class="fi-breadcrumb-item emotion-2"
    >
      <span
        aria-current="page"
        class="emotion-6"
      >
        Sub sub page
      </span>
    </li>
  </ol>
</nav>
`;
